### PR TITLE
Make /metrics listen on a different address

### PR DESCRIPTION
**Pull request by reynico**
- [x] read the [CONTRIBUTING guidelines](README.md#user-content-contributing)
- [x] raised a GitHub issue or discussed it on the projects chat beforehand
- [x] added unit tests
- [x] added integration tests
- [x] updated documentation if needed
- [x] updated CHANGELOG.md

Addresses #343

### CODE REVIEW

**Comment by kradalby** [comment]
It seems a bit excessive to allocate a different port for this, limiting a network would seem sufficient? This would break peoples current setup.

**Comment by reynico** [comment]
> limiting a network would seem sufficient?
not sure if there's any way to do it per URI. I've found https://github.com/bu/gin-access-limit but it needs to be attached to the whole Gin engine. 

> This would break peoples current setup.
I could add a switch so if there's no `metrics_listen_addr` it will use `listen_addr` so it'll be backward compatible. What do you think?

**Comment by kradalby** [comment]
Sorry for the late reply, I have not had time to make a decision here, I do not really care for the behaviour of having a flat set and not set.

It might be ok to have a separate listen, but it will have to wait a couple of releases, I'll get back to you when we have finished other stuff that are a bit pressing.

**Comment by ohdearaugustin** [comment]
I would maybe also add a option to enable/disable the metrics endpoint.  Furthermore as it is a prometheus exporter I would suggest to use a port in the default port range of prometheus exporters. [default port allocations](https://github.com/prometheus/prometheus/wiki/Default-port-allocations)

**Comment by reynico** [comment]
@ohdearaugustin done! I picked `9915` as the Prometheus port and also added a switch to toggle it.


btw, seems like some of the tests on upstream are broken.

**Comment by kradalby** [comment]
So remove the enable flag, and set default listen to local:9090.

**Comment by ohdearaugustin** [comment]
@kradalby Those metrics exported by gin are prometheus metrics. So I don't understand the argumentation that this is not an exporter. It is an exporter integrated in the application. For example Gitlab does the same with it's "exporters" they are also integrated into the application itself.

```
The ports in the linked list is to be reserved for applications specifically designed for translating api to metrics.
```

Never heard of that limitation beforehand. And furthermore why would you remove the ability to enable/disable the endpoint?

**Comment by kradalby** [comment]
> Those metrics exported by gin are prometheus metrics. So I don't understand the argumentation that this is not an exporter. 

And exporter is a component _translating_ information that is not Prometheus compatible to Prometheus metrics[1]. For example, by querying an API and exporting Metrics. 

What we do is instrument the code directly, and not rely on a thirdparty/microservice outside of `headscale`, and it is by the quote posted _not_ an exporter:

> ```
> The ports in the linked list is to be reserved for applications specifically designed for translating api to metrics.
> ```

The "correct" or optimal way to use prometheus, is for all applications in the world to just have a `/metrics` endpoint, making exporters unnecessary. We have built it in, but that page of ports is for apps that yet not have built-ins, e.g. Postgres. or where it makes sense to keep it outside from a performance perspective. 


> It is an exporter integrated in the application. For example Gitlab does the same with it's "exporters" they are also integrated into the application itself.

Basically GitLab is an example **breaking** the rule.

> Never heard of that limitation beforehand. And furthermore why would you remove the ability to enable/disable the endpoint?

Metrics should always be available and it also serves as a health endpoint.


1. https://prometheus.io/docs/instrumenting/exporters/ describes a bit the difference of exporter and instrumentations.

**Comment by ohdearaugustin** [comment]
@kradalby Thank you for that information and you convinced me. I was wrong.
@reynico Sorry for also confusing you.

**Comment by kradalby** [comment]
No problem ;) Most of my days are spent in Prometheus land, so I got this into my head now :P

**Comment by reynico** [comment]
No prob! I'll revert the last two commits, plus set the listen port to `9090`.

**Comment by kradalby** [comment]
Looks sensible, could you add an entry to the CHANGELOG under breaking?

**Comment by reynico** [comment]
Done @kradalby ! Want me to add something at features/changes as well?

**Review by kradalby — APPROVED** [review]
